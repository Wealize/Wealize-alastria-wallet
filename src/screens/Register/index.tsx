import React, { useState, useEffect } from 'react'
import { NavigationFunctionComponent } from 'react-native-navigation'

import { createEncryptedWallet } from '../../services/WalletService'
import LocalStorageService, {
  STORAGE_KEYS
} from '../../services/LocalStorageService'
import { NavigationProps, setStackRoot } from '../../utils/navigation-utils'
import {
  ContainerBottom,
  InputsContainer,
  Subtitle,
  Title,
  MainContainer,
  TermsCheckBox,
  InputStyled,
  TermsText,
  SplashActivityIndicator
} from './styles'
import { REGISTER } from '../../constants/text'
import { Colors } from '../../utils/themes'
import TouchableButton from '../../components/TouchableButton'
import { SCREEN } from '../../constants/screens'

const Register: NavigationFunctionComponent = ({
  componentId
}: NavigationProps) => {
  const [isPinFilled, setIsPinFilled] = useState(true)
  const [firstPin, setFirstPin] = useState('')
  const [secondPin, setSecondPin] = useState('')
  const [isLoading, setIsLoading] = useState(false)
  const [isTermsCheckboxChecked, setIsTermsCheckboxChecked] = useState(false)

  useEffect(() => {
    arePinEquals() ? setIsPinFilled(false) : setIsPinFilled(true)
  }, [firstPin, secondPin])

  useEffect(() => {
    if (isLoading) {
      createWallet()
    }
  }, [isLoading])

  const arePinEquals = () => {
    return firstPin !== '' && secondPin !== '' && firstPin === secondPin
  }

  const createWallet = async () => {
    await createEncryptedWallet(firstPin)
    await LocalStorageService.storeBool(STORAGE_KEYS.IS_WALLET_CREATED, true)

    setIsLoading(false)
    setStackRoot(componentId, SCREEN.ALASTRIA_TOKEN_QR_READER)
  }

  const renderErrorMessage = () => {
    if (firstPin === '' || secondPin === '' || arePinEquals()) {
      return ''
    }
    return REGISTER.ERROR
  }

  return (
    <MainContainer>
      <Title>{REGISTER.TITLE}</Title>
      <Subtitle>{REGISTER.SUBTITLE}</Subtitle>

      <InputsContainer>
        <InputStyled
          value={firstPin}
          keyboardType="decimal-pad"
          onChangeText={(value) => setFirstPin(value.replace(/[^0-9]/g, ''))}
          placeholder={REGISTER.PIN}
        />
        <InputStyled
          value={secondPin}
          keyboardType="decimal-pad"
          onChangeText={(value) => setSecondPin(value.replace(/[^0-9]/g, ''))}
          placeholder={REGISTER.SECOND_PIN}
          errorMessage={renderErrorMessage()}
        />
        <TermsCheckBox
          title={
            <TermsText
              onPress={() =>
                setStackRoot(componentId, SCREEN.TERMS_AND_CONDITIONS, {
                  lastComponent: SCREEN.REGISTER
                })
              }
            >
              {REGISTER.CONDITIONS}
            </TermsText>
          }
          checked={isTermsCheckboxChecked}
          onPress={() => {
            setIsTermsCheckboxChecked(!isTermsCheckboxChecked)
          }}
        />
      </InputsContainer>

      <ContainerBottom>
        {isLoading ? (
          <SplashActivityIndicator size={35} color={Colors.indicatorLoading} />
        ) : null}
        <TouchableButton
          title={isLoading ? REGISTER.LOADING : REGISTER.BUTTON}
          disabled={isPinFilled || !isTermsCheckboxChecked || isLoading}
          onPress={() => {
            setIsLoading(true)
          }}
        />
      </ContainerBottom>
    </MainContainer>
  )
}

Register.options = {
  statusBar: {
    backgroundColor: Colors.mainBackground,
    style: 'dark'
  },
  topBar: {
    visible: false,
    elevation: 0, // Remove border in Android
    borderColor: Colors.mainBackground // Remove border in iOS
  },
  animations: {
    push: {
      content: {
        translationX: {
          from: require('react-native').Dimensions.get('window').width,
          to: 0,
          duration: 300
        }
      }
    }
  }
}

export default Register
