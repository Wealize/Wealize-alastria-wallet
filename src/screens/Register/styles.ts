import { Text, View } from 'react-native'
import styled from 'styled-components'
import { UIActivityIndicator } from 'react-native-indicators'
import { CheckBox, Input } from 'react-native-elements'

import ViewContainer from '../../components/ViewContainer'
import { Colors } from '../../utils/themes'

export const MainContainer = styled(ViewContainer)`
  background-color: #ffffff;
  padding-left: 0;
  padding-right: 0;
`

export const Title = styled(Text)`
  font-size: 20px;
  font-weight: bold;
  padding-top: 15%;
  margin-left: 9%;
`

export const Subtitle = styled(Text)`
  margin-top: 13%;
  margin-bottom: 16px;
  margin-left: 9%;
  margin-right: 7%;
`

export const InputsContainer = styled(View)`
  margin-left: 6%;
  margin-right: 8%;
`

export const InputStyled = styled(Input).attrs({
  selectionColor: Colors.activeInput,
  errorStyle: {
    color: Colors.error
  }
})``

export const TermsCheckBox = styled(CheckBox).attrs({
  containerStyle: {
    backgroundColor: Colors.mainBackground,
    borderWidth: 0,
    height: 20,
    marginLeft: 1,
    paddingTop: 8
  },
  size: 24,
  checkedColor: Colors.activeInput,
  uncheckedIcon: 'square-o',
  checkedIcon: 'check-square'
})``

export const TermsText = styled(Text)`
  font-size: 14px;
  color: ${Colors.activeInput};
  font-weight: normal;
  margin-bottom: 1%;
  margin-left: 2%;
`

export const SplashActivityIndicator = styled(UIActivityIndicator)`
  flex: 1;
  justify-content: flex-end;
  align-items: center;
  margin-bottom: 6%;
`

export const ContainerBottom = styled(View)`
  flex: 1;
  justify-content: flex-end;
  align-items: center;
  margin: 20px auto 40px auto;
  width: 100%;
`
