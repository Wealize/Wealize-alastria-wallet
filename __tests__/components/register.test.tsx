import React from 'react'
import renderer from 'react-test-renderer'

import Register from '../../src/screens/Register'


describe('OnboardginComponent', () => {
  it('should render correctly', () => {
    const result = renderer.create(<Register componentId={'1'} />)
    expect(result).toMatchSnapshot()
  })
})