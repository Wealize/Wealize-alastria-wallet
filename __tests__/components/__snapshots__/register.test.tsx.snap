// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardginComponent should render correctly 1`] = `
<View
  className="sc-bdvvtL sc-gsDKAQ dMYQCW gCheCb"
>
  <Text
    className="sc-dkPtRN hHtZNm"
  >
    Registro
  </Text>
  <Text
    className="sc-hKwDye kSCOVw"
  >
    Añade seguridad a tu aplicación creando un PIN de seguridad
  </Text>
  <View
    className="sc-eCImPb ktYsMA"
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          className="sc-jRQBWg"
          editable={true}
          keyboardType="decimal-pad"
          onChangeText={[Function]}
          placeholder="Pin de Seguridad"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          selectionColor="#f06078"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#EE3F61",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          className="sc-jRQBWg"
          editable={true}
          keyboardType="decimal-pad"
          onChangeText={[Function]}
          placeholder="Repetir Pin"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          selectionColor="#f06078"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#EE3F61",
            "fontSize": 12,
            "margin": 5,
          }
        }
      >
        
      </Text>
    </View>
    <View
      className="sc-gKclnd hKiCPJ"
    >
      <View
        accessibilityRole="checkbox"
        accessibilityState={
          Object {
            "checked": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "margin": 5,
            "marginLeft": 10,
            "marginRight": 10,
            "opacity": 1,
            "padding": 5,
          }
        }
        testID="checkbox"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#bfbfbf",
                  "fontSize": 24,
                },
                Object {
                  "minWidth": 24,
                },
                Object {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
      <Text
        className="sc-furwcr heARiX"
        onPress={[Function]}
      >
        Aceptar términos y condiciones
      </Text>
    </View>
  </View>
  <View
    className="sc-jrQzAO jJueht"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#CCCCCC",
            "borderRadius": 45,
            "height": 39,
            "justifyContent": "center",
            "width": 224,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 16,
            }
          }
        >
          Comenzar
        </Text>
      </View>
    </View>
  </View>
</View>
`;
